### PR TITLE
Vapp template lease handling

### PR DESCRIPTION
**Pull request by dataclouder**
Add lease handling to `vcd_catalog_vapp_template`

### CODE REVIEW

File: website/docs/r/catalog_vapp_template.html.markdown [hunk 1]
```diff
@@ -62,6 +62,9 @@ The following arguments are supported:
 * `upload_piece_size` - (Optional) - Size in MB for splitting upload size. It can possibly impact upload performance. Default 1MB
 * `metadata` -  (Deprecated) Use `metadata_entry` instead. Key/value map of metadata to assign to the associated vApp Template
 * `metadata_entry` - (Optional; *v3.8+*) A set of metadata entries to assign. See [Metadata](#metadata) section for details.
+* `lease` - (Optional *v3.11+*) the information about the vApp Template lease. It includes the field below. When this section is
+  included, the field is mandatory. If lease value is higher than the one allowed for the whole Org, we get an error
```
Comment (lvirbalas) [inline website/docs/r/catalog_vapp_template.html.markdown:66]: Nit, for consistency:

```suggestion
* `lease` - (Optional *v3.11+*) The information about the vApp Template lease. It includes the field below. When this section is
  included, the field is mandatory. If lease value is higher than the one allowed for the whole Org, we get an error.
```

File: vcd/resource_vcd_catalog_vapp_template.go [hunk 2]
```diff
@@ -256,6 +294,35 @@ func resourceVcdCatalogVappTemplateDelete(_ context.Context, d *schema.ResourceD
...
+		// No lease block: we read the lease defaults from the Org
+		adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+		if err != nil {
+			return fmt.Errorf("error retrieving admin Org from parent Org in vApp Template %s: %s", vAppTemplate.VAppTemplate.Name, err)
+		}
+		if adminOrg.AdminOrg.OrgSettings == nil || adminOrg.AdminOrg.OrgSettings.OrgVAppTemplateSettings == nil {
+			return fmt.Errorf("error retrieving Org lease settings")
+		}
+		storageLease = *adminOrg.AdminOrg.OrgSettings.OrgVAppTemplateSettings.StorageLeaseSeconds
+	}
```
Comment (adambarreiro) [inline vcd/resource_vcd_catalog_vapp_template.go:315]: I have a doubt about this `else` part. Before this PR introduced the `lease` block, we did nothing in terms of lease. Now, if `lease` is not present, we take Org's value and use it.
Does it mean that previous code was wrong due to lacking of this feature? Should this be done by VCD and it is not? Or it is just custom logic that we are introducing?

**Review by adambarreiro — COMMENTED** [review]
Just a small doubt

File: vcd/resource_vcd_catalog_vapp_template.go (hunk 2 above)
Comment (dataclouder) [reply vcd/resource_vcd_catalog_vapp_template.go:315]: Before, VCD applied the Org defaults without asking.
Now, we have an explicit lease, and we need to match it to the Org defaults to make the plan succeed.
The default would work just as well without it, but we need to take care of the hidden values in the state.

Comment (adambarreiro) [reply vcd/resource_vcd_catalog_vapp_template.go:315]: > Before, VCD applied the Org defaults without asking.
Then we don't need the `else`, do we? VCD will apply the defaults, then they will be read back when Read comes:
```
leaseInfo, err := vAppTemplate.GetLease()
...
err = d.Set("lease", leaseData)
```
Actually I did a try by removing the `else` contents, and not setting any `lease` block, and the state was correctly read:

```
            "lease": [
              {
                "storage_lease_in_sec": 64800
              }
            ],
```
(I configured the Org to have 18h lease)

File: website/docs/d/catalog_vapp_template.html.markdown [hunk 3]
```diff
@@ -118,6 +118,11 @@ The following arguments are supported:
 * `metadata` - (Deprecated) Use `metadata_entry` instead. Key/value map of metadata for the associated vApp template.
 * `metadata_entry` - A set of metadata entries assigned to this vApp Template. See [Metadata](/providers/vmware/vcd/latest/docs/resources/catalog_vapp_template#metadata) section for details.
 
+* `lease` - (*v3.11+*) - The information about the vApp Template lease. It includes the following field:
+  * `storage_lease_in_sec` - How long the vApp Template is available before being automatically deleted or marked as expired. 0 means never expires (or maximum allowed by parent Org allows).
+
+
+
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/d/catalog_vapp_template.html.markdown]: removed

File: website/docs/r/catalog_vapp_template.html.markdown (hunk 1 above)
Comment (dataclouder) [reply website/docs/r/catalog_vapp_template.html.markdown]: Fixed

File: vcd/resource_vcd_catalog_vapp_template.go (hunk 2 above)
Comment (dataclouder) [reply vcd/resource_vcd_catalog_vapp_template.go]: OK. It seems that I overthought the operation. That `else` was not needed, and I removed it.
Thanks for checking it

**Review by adambarreiro — APPROVED** [review]
LGTM!

File: website/docs/r/catalog_vapp_template.html.markdown [hunk 4]
```diff
@@ -62,6 +62,9 @@ The following arguments are supported:
 * `upload_piece_size` - (Optional) - Size in MB for splitting upload size. It can possibly impact upload performance. Default 1MB
 * `metadata` -  (Deprecated) Use `metadata_entry` instead. Key/value map of metadata to assign to the associated vApp Template
 * `metadata_entry` - (Optional; *v3.8+*) A set of metadata entries to assign. See [Metadata](#metadata) section for details.
+* `lease` - (Optional *v3.11+*) The information about the vApp Template lease. It includes the field below. When this section is
+  included, the field is mandatory. If lease value is higher than the one allowed for the whole Org, we get an error
+   * `storage_lease_in_sec` - How long the vApp Template is available before being automatically deleted or marked as expired. 0 means never expires (or maximum allowed by parent Org). Regular values accepted from 3600+.
```
Comment (Didainius) [inline website/docs/r/catalog_vapp_template.html.markdown:67]: * Could we add this section to the existing example snippet so that it is clearer?

**Review by Didainius — APPROVED** [review]

**Review by adezxc — APPROVED** [review]

**Review by lvirbalas — APPROVED** [review]

File: website/docs/r/catalog_vapp_template.html.markdown (hunk 4 above)
Comment (dataclouder) [reply website/docs/r/catalog_vapp_template.html.markdown:71]: done

**Pull request merged**